File: src/Service/BoutAnalysisService.php
<?php
namespace App\Service;

use App\Entity\FightResult;
use App\Entity\FightStatistic;

/**
 * BOUT ANALYSIS ENGINE v2.0 - Senior Implementation
 * 
 * Provides professional-grade statistical modeling for boxing matches,
 * calculating dominance scores, efficiency indices, and tactical insights.
 */
class BoutAnalysisService
{
    /**
     * Executes a comprehensive multi-vector analysis of the bout statistics.
     */
    public function analyzeBout(FightResult $result, array $aggregates, array $allStats): array
    {
        if (count($aggregates) < 2) {
            return $this->buildEmptyAnalysis();
        }

        $s1 = $aggregates[0];
        $s2 = $aggregates[1];
        
        $f1 = $s1->getFighter();
        $f2 = $s2->getFighter();

        // ─── CORE METRICS ───────────────────────────────────────────────────
        $totalLanded = $s1->getPunchesLanded() + $s2->getPunchesLanded();
        $totalThrown = $s1->getPunchesThrown() + $s2->getPunchesThrown();
        $totalKds    = $s1->getKnockdowns() + $s2->getKnockdowns();
        $overallAcc  = $totalThrown > 0 ? ($totalLanded / $totalThrown) * 100 : 0;

        // ─── ADVANCED MODELING ──────────────────────────────────────────────
        $winnerStat = null;
        $domScore = 5.0; // Baseline parity

        if (!$result->isDraw() && $result->getWinner()) {
            $isF1Winner = ($result->getWinner()->getFighterId() === $f1->getFighterId());
            $winnerStat = $isF1Winner ? $s1 : $s2;
            $loserStat  = $isF1Winner ? $s2 : $s1;
            
            // Efficiency Index (Landed Ratio)
            $efficiencyIndex = $loserStat->getPunchesLanded() > 0 ? 
                ($winnerStat->getPunchesLanded() / $loserStat->getPunchesLanded()) : 2.5;
            
            // Tactical Components
            $kdVector  = $winnerStat->getKnockdowns() * 2.0;
            $accVector = ($winnerStat->getPunchAccuracy() - $loserStat->getPunchAccuracy()) * 0.1;
            $volVector = ($winnerStat->getPunchesThrown() > $loserStat->getPunchesThrown()) ? 0.5 : -0.5;
            
            $domScore = min(10.0, 5.0 + ($efficiencyIndex * 1.5) + $kdVector + $accVector + $volVector);
        }

        // ─── NARRATIVE CONSTRUCTION ──────────────────────────────────────────
        $analysis = [];
        $insights = [];

        // 1. Tactical Conclusion
        if ($result->isDraw()) {
            $analysis[] = "A grueling stalemate defined by extreme statistical parity. Neither combatant could definitively solve the other's defensive structure, resulting in a razor-close deadlock.";
        } else {
            $winner = $result->getWinner();
            $method = $result->getMethodOfVictory();
            
            if ($method === FightResult::METHOD_KO) {
                $analysis[] = "A clinical stoppage victory for {$winner->getLastName()}, who translated statistical pressure into a terminal finish in Round {$result->getRoundNumber()}.";
                $insights[] = "{$winner->getLastName()} demonstrated elite 'Stop-Start' kinetic efficiency, concluding the bout before it reached the judges.";
            } else {
                $type = $result->getDecisionType() ?: 'Decision';
                $analysis[] = "A strategic masterclass by {$winner->getLastName()}, utilizing superior ring generalship to secure a {$type} over the distance.";
                $insights[] = "Tactical discipline and range management were the primary differentiators for the victor.";
            }
        }

        // 2. Efficiency & Output Analysis
        $higherVol = ($s1->getPunchesThrown() > $s2->getPunchesThrown()) ? $s1 : $s2;
        $higherAcc = ($s1->getPunchAccuracy() > $s2->getPunchAccuracy()) ? $s1 : $s2;

        $analysis[] = "{$higherVol->getFighter()->getLastName()} dictated the tempo with {$higherVol->getPunchesThrown()} attempts, while {$higherAcc->getFighter()->getLastName()} provided the clinical counter-balance, landing with {$higherAcc->getPunchAccuracy()}% precision.";

        // 3. Kinetic Trends (Round-by-Round)
        $roundsWonByF1 = 0;
        $roundsWonByF2 = 0;
        $lateSurge = ['fighter' => null, 'count' => 0];

        $roundMap = [];
        foreach ($allStats as $st) { $roundMap[$st->getRound()][] = $st; }

        foreach ($roundMap as $r => $stats) {
            if (count($stats) === 2) {
                $r1 = $stats[0]; $r2 = $stats[1];
                $winnerId = ($r1->getPunchesLanded() >= $r2->getPunchesLanded()) ? $r1->getFighter()->getFighterId() : $r2->getFighter()->getFighterId();
                
                if ($winnerId === $f1->getFighterId()) $roundsWonByF1++; else $roundsWonByF2++;
                
                if ($r > 6) {
                    $rWinner = ($winnerId === $f1->getFighterId()) ? $f1 : $f2;
                    if ($lateSurge['fighter'] === $rWinner) $lateSurge['count']++;
                    else { $lateSurge['fighter'] = $rWinner; $lateSurge['count'] = 1; }
                }
            }
        }

        // 4. Detailed Tactical Insights
        $this->generateTacticalInsights($insights, $f1, $f2, $s1, $s2, $roundsWonByF1, $roundsWonByF2, $lateSurge);

        return [
            'summary' => implode(" ", $analysis),
            'dominanceScore' => round($domScore, 1),
            'totalPunchesLanded' => $totalLanded,
            'knockdowns' => $totalKds,
            'overallAccuracy' => round($overallAcc, 1),
            'insights' => $insights
        ];
    }

    private function generateTacticalInsights(array &$insights, $f1, $f2, $s1, $s2, $r1, $r2, $surge): void
    {
        // Connectivity Dominance
        if (abs($r1 - $r2) > 3) {
            $dom = $r1 > $r2 ? $f1 : $f2;
            $insights[] = "{$dom->getLastName()} controlled the connective architecture, outlanding the opponent in " . max($r1, $r2) . " distinct rounds.";
        }

        // Championship Conditioning
        if ($surge['count'] >= 3 && $surge['fighter']) {
            $insights[] = "{$surge['fighter']->getLastName()} demonstrated elite cardiovascular reserves, seizing total kinetic control during the championship rounds.";
        }

        // Power Distribution
        $p1 = $s1->getPowerPunchesLanded(); $p2 = $s2->getPowerPunchesLanded();
        if (abs($p1 - $p2) > 12) {
            $pDom = $p1 > $p2 ? $f1 : $f2;
            $pCount = ($pDom === $f1) ? $p1 : $p2;
            $insights[] = "{$pDom->getLastName()} dominated the heavy-artillery exchanges, landing {$pCount} significant power shots.";
        }

        // Defensive Efficiency
        $acc1 = $s1->getPunchAccuracy(); $acc2 = $s2->getPunchAccuracy();
        if (abs($acc1 - $acc2) > 15) {
            $eff = $acc1 > $acc2 ? $f1 : $f2;
            $insights[] = "Precision gap identified: {$eff->getLastName()} operated at a much higher efficiency threshold than the opposition.";
        }
    }

    private function buildEmptyAnalysis(): array
    {
        return [
            'summary' => "Insufficient kinetic data for professional modeling.",
            'dominanceScore' => 0,
            'totalPunchesLanded' => 0,
            'knockdowns' => 0,
            'overallAccuracy' => 0,
            'insights' => ["Data integrity verification required for full tactical breakdown."]
        ];
    }
}

File: src/main/java/tn/smartfight/service/BoutAnalysisService.java
package tn.smartfight.service;

import tn.smartfight.model.FightStatistic;

import java.util.*;

/**
 * Java port of PHP BoutAnalysisService.
 * Provides statistical modeling: dominance score, efficiency index, tactical insights.
 */
public class BoutAnalysisService {

    public static class BoutAnalysis {
        public String summary = "";
        public double dominanceScore = 0;
        public int totalPunchesLanded = 0;
        public int knockdowns = 0;
        public double overallAccuracy = 0;
        public List<String> insights = new ArrayList<>();
    }

    /**
     * @param fightId       the resultId of the fight
     * @param winner        name of winner, null for draw
     * @param method        methodOfVictory string
     * @param roundEnded    round number the fight ended
     * @param decisionType  decision_type string
     * @param agg1          aggregated stats for fighter 1
     * @param agg2          aggregated stats for fighter 2
     * @param allRoundStats all per-round stats (both fighters)
     */
    public BoutAnalysis analyzeBout(
            int fightId,
            String winner,
            String method,
            int roundEnded,
            String decisionType,
            FightStatistic agg1,
            FightStatistic agg2,
            List<FightStatistic> allRoundStats
    ) {
        if (agg1 == null || agg2 == null) return empty();

        int totalLanded = agg1.getPunchesLanded() + agg2.getPunchesLanded();
        int totalThrown = agg1.getPunchesThrown() + agg2.getPunchesThrown();
        int totalKds    = agg1.getKnockdowns() + agg2.getKnockdowns();
        double overallAcc = totalThrown > 0 ? (totalLanded * 100.0 / totalThrown) : 0;

        double domScore = 5.0;
        FightStatistic winnerStat = null;
        FightStatistic loserStat  = null;

        boolean isDraw = (winner == null || winner.isBlank());

        if (!isDraw) {
            boolean isF1Winner = agg1.getFighterName() != null &&
                    agg1.getFighterName().toLowerCase().contains(extractLastName(winner).toLowerCase());
            winnerStat = isF1Winner ? agg1 : agg2;
            loserStat  = isF1Winner ? agg2 : agg1;

            double efficiencyIndex = loserStat.getPunchesLanded() > 0
                    ? (double) winnerStat.getPunchesLanded() / loserStat.getPunchesLanded()
                    : 2.5;
            double kdVector  = winnerStat.getKnockdowns() * 2.0;
            double accVector = (winnerStat.getPunchAccuracy() - loserStat.getPunchAccuracy()) * 0.1;
            double volVector = winnerStat.getPunchesThrown() > loserStat.getPunchesThrown() ? 0.5 : -0.5;
            domScore = Math.min(10.0, 5.0 + efficiencyIndex * 1.5 + kdVector + accVector + volVector);
        }

        List<String> analysis = new ArrayList<>();
        List<String> insights = new ArrayList<>();

        if (isDraw) {
            analysis.add("A grueling stalemate defined by extreme statistical parity. " +
                    "Neither combatant could definitively solve the other's defensive structure, " +
                    "resulting in a razor-close deadlock.");
        } else {
            String wName = extractLastName(winner);
            if ("KO".equalsIgnoreCase(method) || "TKO".equalsIgnoreCase(method)) {
                analysis.add("A clinical stoppage victory for " + wName +
                        ", who translated statistical pressure into a terminal finish in Round " + roundEnded + ".");
                insights.add(wName + " demonstrated elite 'Stop-Start' kinetic efficiency, " +
                        "concluding the bout before it reached the judges.");
            } else {
                String dtype = (decisionType != null && !decisionType.isBlank()) ? decisionType : "Decision";
                analysis.add("A strategic masterclass by " + wName +
                        ", utilizing superior ring generalship to secure a " + dtype + " over the distance.");
                insights.add("Tactical discipline and range management were the primary differentiators for the victor.");
            }
        }

        FightStatistic higherVol = agg1.getPunchesThrown() >= agg2.getPunchesThrown() ? agg1 : agg2;
        FightStatistic higherAcc = agg1.getPunchAccuracy() >= agg2.getPunchAccuracy() ? agg1 : agg2;
        String volName = lastName(higherVol);
        String accName = lastName(higherAcc);
        analysis.add(volName + " dictated the tempo with " + higherVol.getPunchesThrown() +
                " attempts, while " + accName + " provided the clinical counter-balance, landing with " +
                String.format("%.1f", higherAcc.getPunchAccuracy()) + "% precision.");

        int r1Wins = 0, r2Wins = 0;
        Map<Integer, List<FightStatistic>> roundMap = new LinkedHashMap<>();
        for (FightStatistic s : allRoundStats) {
            if (s.getRoundNumber() == null) continue;
            roundMap.computeIfAbsent(s.getRoundNumber(), k -> new ArrayList<>()).add(s);
        }

        FightStatistic lateLeader = null;
        int lateCount = 0;

        for (Map.Entry<Integer, List<FightStatistic>> entry : roundMap.entrySet()) {
            List<FightStatistic> rStats = entry.getValue();
            if (rStats.size() < 2) continue;
            FightStatistic r1 = rStats.get(0);
            FightStatistic r2 = rStats.get(1);
            boolean f1WinsRound = r1.getPunchesLanded() >= r2.getPunchesLanded();
            if (f1WinsRound) r1Wins++; else r2Wins++;
            if (entry.getKey() > 6) {
                FightStatistic rWinner = f1WinsRound ? r1 : r2;
                if (lateLeader != null && lateLeader.getFighterId() == rWinner.getFighterId()) {
                    lateCount++;
                } else {
                    lateLeader = rWinner;
                    lateCount = 1;
                }
            }
        }

        generateTacticalInsights(insights, agg1, agg2, r1Wins, r2Wins, lateLeader, lateCount);

        BoutAnalysis result = new BoutAnalysis();
        result.summary           = String.join(" ", analysis);
        result.dominanceScore    = Math.round(domScore * 10.0) / 10.0;
        result.totalPunchesLanded = totalLanded;
        result.knockdowns        = totalKds;
        result.overallAccuracy   = Math.round(overallAcc * 10.0) / 10.0;
        result.insights          = insights;
        return result;
    }

    private void generateTacticalInsights(List<String> insights,
                                          FightStatistic s1, FightStatistic s2,
                                          int r1Wins, int r2Wins,
                                          FightStatistic lateSurgeLeader, int surgeCount) {
        if (Math.abs(r1Wins - r2Wins) > 3) {
            FightStatistic dom = r1Wins > r2Wins ? s1 : s2;
            insights.add(lastName(dom) + " controlled the connective architecture, outlanding the opponent in " +
                    Math.max(r1Wins, r2Wins) + " distinct rounds.");
        }
        if (surgeCount >= 3 && lateSurgeLeader != null) {
            insights.add(lateSurgeLeader.getFighterName() + " demonstrated elite cardiovascular reserves, " +
                    "seizing total kinetic control during the championship rounds.");
        }
        int p1 = s1.getPowerPunchesLanded(), p2 = s2.getPowerPunchesLanded();
        if (Math.abs(p1 - p2) > 12) {
            FightStatistic pDom = p1 > p2 ? s1 : s2;
            insights.add(lastName(pDom) + " dominated the heavy-artillery exchanges, landing " +
                    Math.max(p1, p2) + " significant power shots.");
        }
        double acc1 = s1.getPunchAccuracy(), acc2 = s2.getPunchAccuracy();
        if (Math.abs(acc1 - acc2) > 15) {
            FightStatistic eff = acc1 > acc2 ? s1 : s2;
            insights.add("Precision gap identified: " + lastName(eff) +
                    " operated at a much higher efficiency threshold than the opposition.");
        }
    }

    private BoutAnalysis empty() {
        BoutAnalysis a = new BoutAnalysis();
        a.summary = "Insufficient kinetic data for professional modeling.";
        a.insights.add("Data integrity verification required for full tactical breakdown.");
        return a;
    }

    private String lastName(FightStatistic s) {
        if (s == null || s.getFighterName() == null) return "Fighter";
        String[] parts = s.getFighterName().trim().split("\\s+");
        return parts.length > 1 ? parts[parts.length - 1] : parts[0];
    }

    private String extractLastName(String fullName) {
        if (fullName == null || fullName.isBlank()) return "";
        String[] parts = fullName.trim().split("\\s+");
        return parts.length > 1 ? parts[parts.length - 1] : parts[0];
    }
}
